fdbserver/ptxn/TLogServer.actor.cpp

```diff
@@ -2113,6 +2134,119 @@ void updatePersistentPopped(Reference<TLogGroupData> self,
 	}
 }
 
+ACTOR Future<Void> updatePoppedLocation(Reference<TLogGroupData> self,
+                                        Reference<LogGenerationData> logData,
+                                        Reference<LogGenerationData::StorageTeamData> data) {
+	// For anything spilled by value, we do not need to track its popped location.
+	if (logData->shouldSpillByValue(data->storageTeamId)) {
+		return Void();
+	}
+
+	if (data->versionForPoppedLocation >= data->persistentPopped) // both are 0 here, thus not continuing.
+		return Void();
+	data->versionForPoppedLocation = data->persistentPopped;
+
+	// Use persistentPopped and not popped, so that a pop update received after spilling doesn't cause
+	// us to remove data that still is pointed to by SpilledData in the btree.
+	if (data->persistentPopped <= logData->persistentDataVersion) {
+		// Recover the next needed location in the Disk Queue from the index.
+		RangeResult kvrefs = wait(self->persistentData->readRange(
+		    KeyRangeRef(persistStorageTeamMessageRefsKey(logData->logId, data->storageTeamId, data->persistentPopped),
+		                persistStorageTeamMessageRefsKey(
+		                    logData->logId, data->storageTeamId, logData->persistentDataVersion + 1)),
+		    1));
+
+		if (kvrefs.empty()) {
+			// Nothing was persistent after all.
+			data->nothingPersistent = true;
+		} else {
+			VectorRef<SpilledData> spilledData;
+			BinaryReader r(kvrefs[0].value, AssumeVersion(logData->protocolVersion));
+			r >> spilledData;
+
+			for (const SpilledData& sd : spilledData) {
+				if (sd.version >= data->persistentPopped) {
+					data->poppedLocation =
+					    sd.start; // stoargeTeamData updates `poppedLocation` through data from spilling
+					data->versionForPoppedLocation = sd.version;
+					break;
+				}
+			}
+		}
+	}
+
+	if (data->persistentPopped >= logData->persistentDataVersion || data->nothingPersistent) {
+		// Then the location must be in memory.
+		auto locationIter = logData->versionLocation.lower_bound(data->persistentPopped);
+		if (locationIter != logData->versionLocation.end()) {
+			data->poppedLocation = locationIter->value.first;
+			data->versionForPoppedLocation = locationIter->key;
+		} else {
+			// No data on disk and no data in RAM.
+			// This TLog instance will be removed soon anyway, so we temporarily freeze our poppedLocation
+			// to avoid trying to track what the ending location of this TLog instance was.
+		}
+	}
+
+	return Void();
+}
+
+// It runs against the oldest TLog instance, calculates the first location in the disk queue that contains un-popped
```

Comment:
Is this still true? Will there be multiple instances of the same tlog group?

Reply:
will check and fix if necessary, did not carefully read